File: Find_numbers_disappeared_in_array.py
#Time Compexity: O(n) its O(2n) because the given array is traversed twice
#Space Complexity: O(1) because no other data structure is used only the input(given) array is used
class Solution():
    def findDisappearedNumbers(self, nums):
        #sanity check
        if (len(nums)==0 or nums==None):
            return None
        #get length of the array
        n = len(nums)
        for i in (range(0,n)):
            #store the value in a variable
            val = abs(nums[i])
            #if the number is +ve
            if nums[val-1] > 0:
                #make the number at the index-1 as -ve
                nums[val-1] = nums[val-1]*(-1)
        result=[]
        for j in (range(0,n)):
            if nums[j] > 0:
                result.append(j+1)
        return result

File: FindNumbersDisappearedInArray.java

// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :yes


// Your code here along with comments explaining your approach
/*Approach

1) calculate the index and calculate the value.
2) if number at index >0
    then the number is at -1;
3) if the number at index is positive 
    -> then the number at index should be added in array List in result.

*/


class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        
        List<Integer> res = new ArrayList();
        
        for(int i=0;i<nums.length;i++)
        {
            int index= Math.abs(nums[i]);
            if(nums[index-1]>0)
            {
                nums[index-1]*=-1;
            }
            
        }
        
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]>0)
            {
                res.add(i+1);
            }
        }
        
        return res;
    }
}